### PR TITLE
Port more of the Mojolicious Growing guide to mojo.js

### PR DESCRIPTION
**Pull request by marcusramberg**
'Substitute 'damn' every time you're inclined to write 'very'; your
 editor will delete it and the writing will be just as it should be.'
  -- Mark Twain

### CODE REVIEW

File: docs/Introduction.md [hunk 1]
```diff
@@ -87,6 +87,8 @@ app.start();
 There is also a helper command available to generate a small example application for you.
 
 ```
+$ mkdir myapp && cd myapp
+$ npm i @mojojs/core
```
Comment (kraih) [inline docs/Introduction.md:91]: I think we already discussed this in the other PR. 🙄

File: docs/Growing.md [hunk 2]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+full-fledged web framework in single-file mode, and as a micro web framework optimized for rapid prototyping.
+
+### Differences
+
+You likely know the feeling, you've got a really cool idea and want to try it as quickly as possible. That's exactly
+why mojo.js apps can be run as a single file (plus package.json for deps).
+
+```
+package.json // Will be generated when you install mojo
+myapp.js     // templates can be inlined in the file.
```
Comment (kraih) [inline docs/Growing.md:132]: "templates" is lower case, and the distance between code and comments is usually 3 spaces.

File: docs/Growing.md [hunk 3]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+|- controllers/                       // Controller directory
+|   |- example.js                     // Controller containing route actions
+|- models/                            // Model directory
+|- public/                            // Static file directory (served automatically)
+|   |- index.html                     // Static HTML file
+|- views/                             // Views directory
+|   |- example/                       // View directory for "Example"
+|   |   |- welcome.html.ejs           // EJS template generating HTML
+|   |- layouts/                       // View directory for layouts
+|   |   |- default.html.ejs           // Layout view/template
```
Comment (kraih) [inline docs/Growing.md:157]: 3 spaces distance to comments.

File: docs/Growing.md [hunk 4]
```diff
@@ -72,13 +72,13 @@ and all session state is kept client-side.
 +---------+                        +------------+
 ```
 
-While HTTP methods such as `PUT`, `GET` and `DELETE` are not directly part of REST they go very well with it and are
+While HTTP methods such as `PUT`, `GET` and `DELETE` are not directly part of REST they go well with it and are
```
Comment (kraih) [inline docs/Growing.md:75]: The original says "very". Why did you change it?

File: docs/Growing.md [hunk 5]
```diff
@@ -72,13 +72,13 @@ and all session state is kept client-side.
...
 commonly used to manipulate resources.
 
 ### Sessions
 
 HTTP was designed as a stateless protocol, web servers don't know anything about previous requests, which makes
-user-friendly login systems very tricky. Sessions solve this problem by allowing web applications to keep stateful
+user-friendly login systems tricky. Sessions solve this problem by allowing web applications to keep stateful
```
Comment (kraih) [inline docs/Growing.md:81]: Why change the original?

File: docs/Growing.md [hunk 6]
```diff
@@ -100,7 +100,7 @@ Content-Length: 16
 Hello again sebastian.
 ```
 
-Traditionally all session data was stored on the server-side and only session ids were exchanged between browser and
+Traditionally all session data was stored on the server-side and only session IDs were exchanged between browser and
```
Comment (kraih) [inline docs/Growing.md:103]: "ids" is lower case in the original.

Comment (kraih) [reply docs/Growing.md:103]: Why make these arbitrary changes from the original growing guide? You didn't open a matching PR for it.

File: docs/Growing.md [hunk 7]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
 having good test coverage and code being designed for testability, which will in turn often prevent future changes from
 breaking old code. Much of mojo.js was developed using TDD.
 
+## Prototype
+
+An important difference between mojo.js and other web frameworks is that it can operate in two modes, both as a 
+full-fledged web framework in single-file mode, and as a micro web framework optimized for rapid prototyping.
```
Comment (kraih) [inline docs/Growing.md:123]: I find the sentence `...both as a  full-fledged web framework in single-file mode, and as a micro web framework optimized for rapid prototyping.` rather confusing. What even is a ` full-fledged web framework in single-file mode`?

File: docs/Growing.md [hunk 8]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+
+### Differences
+
+You likely know the feeling, you've got a really cool idea and want to try it as quickly as possible. That's exactly
+why mojo.js apps can be run as a single file (plus package.json for deps).
```
Comment (kraih) [inline docs/Growing.md:128]: The `...can be run as a single file (plus package.json for deps)` can probably be phrased better.

File: docs/Growing.md [hunk 9]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+
+```
+package.json // Will be generated when you install mojo
+myapp.js     // templates can be inlined in the file.
+```
+
+Full mojo.js applications on the other hand are much closer to a well organized NPM distribution to maximize
+maintainability.
```
Comment (kraih) [inline docs/Growing.md:136]: This comparison doesn't really work, with the `package.json` the minimal version is a proper NPM distribution too.

File: docs/Growing.md [hunk 10]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+|- public/                            // Static file directory (served automatically)
+|   |- index.html                     // Static HTML file
+|- views/                             // Views directory
+|   |- example/                       // View directory for "Example"
+|   |   |- welcome.html.ejs           // EJS template generating HTML
+|   |- layouts/                       // View directory for layouts
+|   |   |- default.html.ejs           // Layout view/template
+```
+
+These application skeletons can be automatically generated with the mojo commands `generate-lite-app` and
```
Comment (kraih) [inline docs/Growing.md:160]: There is no `generate-lite-app` command.

Comment (kraih) [reply docs/Growing.md:160]: The original start `Both application skeletons...` sounds better to me.

File: docs/Growing.md [hunk 11]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+`generate-full-app` - in both cases you need to create a folder for the skeleton and run `npm i @mojojs/core`
```
Comment (kraih) [inline docs/Growing.md:161]: You create a folder for the application, not the skeleton.

File: docs/Growing.md [hunk 12]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+first to install mojolicious for your new app.
+
+```
+$ mkdir myapp && cd myapp
+$ npm i @mojojs/core
+$ npx mojo generate-full-app myapp # or
+$ npx mojo generate-lite-app myapp
+```
+
+Featurewise both are almost equal, the only real differences are organizational, so each one can be gradually
```
Comment (kraih) [inline docs/Growing.md:171]: "Featurewise" is a typo.

File: docs/Growing.md (hunk 4 above)
Comment (marcusramberg) [reply docs/Growing.md:75]: Because of the quote in the commit message (My text linter complains about the use of very).

File: docs/Growing.md [hunk 13]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+const app = mojo();
+
+app.get('/', async ctx => {
+  await ctx.render({text: 'Hello World!'})
+});
+
+app.start();
+```
+
+and press <kbd>Ctrl</kbd> + <kbd>d</kbd>
```
Comment (kraih) [inline docs/Growing.md:200]: What is that?

File: docs/Growing.md [hunk 14]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+
+## Foundation
+
+We start our new application with a single JavaScript file.
+
+```
+$ mkdir login-manager
+$ cd login-manager
+$ npm i @mojojs/core
+$ cat > index.js
```
Comment (kraih) [inline docs/Growing.md:182]: That's not a real thing. Please just stick to the original.

File: docs/Growing.md [hunk 15]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+```
+
+Paste in this
```
Comment (kraih) [inline docs/Growing.md:185]: This is not the workflow of the original guide.

File: docs/Growing.md [hunk 16]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+
+``` perl
```
Comment (kraih) [inline docs/Growing.md:187]: Not Perl.

File: docs/Growing.md [hunk 17]
```diff
@@ -117,6 +117,132 @@ functionality and then moves on to producing code that passes these tests. There
...
+
+This will be the foundation for our login manager example application. Let's start it as a server
+with node:
+
+```
+$ node ./index.js server
+[640563] Web application available at http://0.0.0.0:3000/
+```
+
+Note how it outputs the PID, and the URL you can reach the server on. If you want a bit of convenice, we recommend
```
Comment (kraih) [inline docs/Growing.md:210]: The output format with the PID is not a real feature and will probably not be in the final version.

File: docs/Growing.md (hunk 8 above)
Comment (kraih) [reply docs/Growing.md:128]: Actually the original still sounds good here in context, `...that's exactly
why mojo.js applications don't need more than a single file.`

File: docs/Introduction.md (hunk 1 above)
Comment (marcusramberg) [reply docs/Introduction.md:91]: Ok, personally *I* think repetition and full examples are better in tutorial guides, while reference guides should be concise and DRY. But I'll update to not include the full example since you hate it so much.

File: docs/Growing.md (hunk 6 above)
Comment (marcusramberg) [reply docs/Growing.md:103]: ids is a misspelling, and I'd be happy to open a matching PR for the original guide for this and very. https://english.stackexchange.com/questions/101248/how-should-the-abbreviation-for-identifier-be-capitalized

File: docs/Growing.md (hunk 7 above)
Comment (marcusramberg) [reply docs/Growing.md:123]: I agree, I got the order mixed up there when rewriting the sentence to make sense in mojo.js context. Will push an update.

File: docs/Growing.md (hunk 9 above)
Comment (kraih) [reply docs/Growing.md:136]: `Full mojo.js applications on the other hand follow the MVC pattern more closely and separate concerns into different files to maximize maintainability.`

File: docs/Growing.md (hunk 10 above)
Comment (marcusramberg) [reply docs/Growing.md:160]: It duplicates the use of 'both' in the paragraph, which sounds repetitive.

File: docs/Growing.md (hunk 12 above)
Comment (marcusramberg) [reply docs/Growing.md:171]: Should have known better than trusting Wiktionary. 🤦🏻 Will revert to the original spelling.

File: docs/Growing.md (hunk 4 above)
Comment (kraih) [reply docs/Growing.md:75]: Then the text should be changed in the original first.

File: docs/Growing.md (hunk 6 above)
Comment (kraih) [reply docs/Growing.md:103]: Please do.

File: docs/Growing.md (hunk 14 above)
Comment (marcusramberg) [reply docs/Growing.md:182]: image

File: docs/Growing.md (hunk 17 above)
Comment (kraih) [reply docs/Growing.md:210]: Better just don't mention the PID.

File: docs/Growing.md (hunk 13 above)
Comment (kraih) [reply docs/Growing.md:200]: I guess it doesn't really matter, please just stick with the original workflow, or make a PR to the Perl guide first.

Comment (marcusramberg) [reply docs/Growing.md:200]: it makes the markdown render like this in github markdown:
image

File: docs/Growing.md (hunk 10 above)
Comment (kraih) [reply docs/Growing.md:160]: Here i actually wouldn't mind if you just repeat the `mkdir`/`npm i...` and leave it out of the text description.

Comment (kraih) [reply docs/Growing.md:160]: Since the context is starting fresh with a new app. As opposed the introduction where we already had a directory creation section.

File: docs/Introduction.md (hunk 1 above)
Comment (marcusramberg) [reply docs/Introduction.md]: Oh wait. I didn't even consider I'd included this change in the intro in this branch. Sorry. 🤦🏻

File: docs/Growing.md (hunk 14 above)
Comment (kraih) [reply docs/Growing.md:180]: Even with the picture i'm still not sure what you're trying to show me.

Comment (kraih) [reply docs/Growing.md:180]: Do you want beginners to use `cat` as an editor? 🤨

Comment (kraih) [reply docs/Growing.md:180]: (but i digress... sorry... please just stick with the original workflow)

Comment (marcusramberg) [reply docs/Growing.md:180]: I was just demonstrating that it's a real thing :) But Ok. I'll revert to using touch even if I don't understand the rationale for creating empty files.

Comment (kraih) [reply docs/Growing.md]: Ah, i was just being a bit hyperbolic with the first comment. 😁

File: docs/Growing.md (hunk 4 above)
Comment (marcusramberg) [reply docs/Growing.md:75]: I guess this should be ok now that the change has been applied upstream, will resolve this and the other two similar one.

File: docs/Growing.md [hunk 18]
```diff
@@ -117,6 +117,126 @@ functionality and then moves on to producing code that passes these tests. There
...
+full-fledged web framework, and as a single file micro web framework optimized for rapid prototyping.
+
+### Differences
+
+You likely know the feeling, you've got a really cool idea and want to try it as quickly as possible. That's exactly
+why mojo.js applications don't need more than a single JavaScript file (in addition to package.json).
+
+```
+package.json // Will be generated when you install mojo
+myapp.js     // Templates can be inlined in the file.
```
Comment (kraih) [inline docs/Growing.md:132]: That's still not 3 spaces.

File: docs/Growing.md [hunk 19]
```diff
@@ -117,6 +117,126 @@ functionality and then moves on to producing code that passes these tests. There
...
+An important difference between mojo.js and other web frameworks is that it can operate in two modes, both as a 
+full-fledged web framework, and as a single file micro web framework optimized for rapid prototyping.
+
+### Differences
+
+You likely know the feeling, you've got a really cool idea and want to try it as quickly as possible. That's exactly
+why mojo.js applications don't need more than a single JavaScript file (in addition to package.json).
+
+```
+package.json // Will be generated when you install mojo
```
Comment (kraih) [inline docs/Growing.md:131]: `mojo.js`.

File: docs/Growing.md [hunk 20]
```diff
@@ -117,6 +117,126 @@ functionality and then moves on to producing code that passes these tests. There
...
+```
+$ mkdir login-manager
+$ cd login-manager
+$ npm i @mojojs/core
+$ touch index.js
+```
+
+This will be the foundation for our login manager example application. 
+
+``` javascript
```
Comment (kraih) [inline docs/Growing.md:185]: Think the rest of the docs just use `js`.
